dhis-2/dhis-api/src/main/java/org/hisp/dhis/scheduling/JobConfiguration.java

```diff
@@ -103,101 +109,110 @@ public class JobConfiguration extends BaseIdentifiableObject implements Secondar
-   * @param inMemoryJob whether this is an in-memory job.
+   * @param type of the job to run once
    */
-  public JobConfiguration(String name, JobType jobType, String executedBy, boolean inMemoryJob) {
-    this.name = name;
-    this.jobType = jobType;
-    this.executedBy = executedBy;
-    this.inMemoryJob = inMemoryJob;
-    init();
+  public JobConfiguration(@Nonnull JobType type) {
+    this(null, type);
   }
 
-  /**
-   * Constructor which implies enabled true and in-memory job false.
-   *
-   * @param name the job name.
-   * @param jobType the {@link JobType}.
-   * @param cronExpression the cron expression.
-   * @param jobParameters the job parameters.
-   */
-  public JobConfiguration(
-      String name, JobType jobType, String cronExpression, JobParameters jobParameters) {
-    this(name, jobType, cronExpression, jobParameters, true, false);
+  public JobConfiguration(@CheckForNull String name, @Nonnull JobType type) {
+    this(name, type, null);
   }
 
   /**
-   * Constructor.
+   * Constructor to use for any type of {@link SchedulingType#ONCE_ASAP} execution.
    *
-   * @param name the job name.
-   * @param jobType the {@link JobType}.
-   * @param cronExpression the cron expression.
-   * @param jobParameters the job parameters.
-   * @param enabled whether this job is enabled.
-   * @param inMemoryJob whether this is an in-memory job.
+   * @param name unique name for the job
+   * @param type of the job to run once
+   * @param executedBy UID of the user running the job or null for run as admin
    */
   public JobConfiguration(
-      String name,
-      JobType jobType,
-      String cronExpression,
-      JobParameters jobParameters,
-      boolean enabled,
-      boolean inMemoryJob) {
-    this.name = name;
-    this.cronExpression = cronExpression;
-    this.jobType = jobType;
-    this.jobParameters = jobParameters;
-    this.enabled = enabled;
-    this.inMemoryJob = inMemoryJob;
-    init();
+      @CheckForNull String name, @Nonnull JobType type, @CheckForNull String executedBy) {
+    this.name =
+        name == null || name.isEmpty()
+            ? "%s (%d)".formatted(type.name(), Instant.now().toEpochMilli())
+            : name;
+    this.jobType = type;
+    this.executedBy = executedBy;
+    setAutoFields();
   }
 
   // -------------------------------------------------------------------------
   // Logic
   // -------------------------------------------------------------------------
 
-  private void init() {
-    if (inMemoryJob) {
-      setAutoFields();
-    }
+  public SchedulingType getSchedulingType() {
+    if (schedulingType != null) return schedulingType;
+    if (cronExpression != null) return SchedulingType.CRON;
+    if (delay != null) return SchedulingType.FIXED_DELAY;
+    return SchedulingType.ONCE_ASAP;
+  }
+
+  public JobStatus getJobStatus() {
+    if (jobStatus != null) return jobStatus;
+    if (getSchedulingType() == SchedulingType.ONCE_ASAP) return JobStatus.NOT_STARTED;
```

Comment:
these checks seem like they could be avoided if a correct `jobStatus` is always set on creation depending on its type. Or is there some difficulty with that approach?

Reply:
Yes, this solution was chosen because of the different code paths. In particular the generic code that creates objects via reflection from user input. To not impose any new requirements on the user input while having the reasonable behaviour should this not be given explicitly this was the best solution that keeps the logic in one place.